File: front/src/lib/product.ts
import axios from 'axios';
import { API_URLS } from '@/conf/env';
import apiClient, { apiService } from './api';

// 新增：配料接口定义
export interface Ingredient {
  id: string;
  name: string;
  category: string; // '谷物', '果蔬', '蛋白质', '添加剂', '其他'
  is_allergen: boolean; // 后端字段名可能是 snake_case
  description?: string;
}

// 新增：生产商接口定义
export interface Manufacturer {
  id: string;
  name: string;
  contact_phone?: string;
  address?: string;
  certification_number?: string;
  created_at: string;
}

// 更新：产品接口定义
export interface Product {
  id: string;
  name: string;
  description?: string;
  price: number;
  category: string;
  imageURL?: string;
  status: string; // 'available', 'in_blind_box', 'sold'
  productionDate: string;
  shelfLifeHours: number;
  createdAt: string;
  manufacturerId: string;    // 新增
  batchNumber?: string;       // 新增
  storageCondition: string;  // 新增: '常温', '冷藏', '冷冻'
  ingredients?: Ingredient[]; // 新增: 关联的配料 (详情时可能返回)
  manufacturer?: Manufacturer;// 新增: 关联的生产商 (详情时可能返回)
  creatorName?: string;       // DTO字段
  manufacturerName?: string; // DTO字段 (列表可能返回)
}

// 用于创建/更新产品时传递的数据结构
export interface ProductInputData {
  name: string;
  description?: string;
  price: number;
  category: string;
  imageURL?: string;
  productionDate: string;    // ISO 8601 format string
  shelfLifeHours: number;
  manufacturerId: string;
  batchNumber?: string;
  storageCondition: string; // '常温', '冷藏', '冷冻'
  ingredientIds?: string[];  // 传递配料ID列表
}

export interface ProductListOptions {
  page?: number;
  size?: number;
  category?: string;
  status?: string;
  keyword?: string;
  sortBy?: string;
  order?: string;
  minPrice?: number;
  maxPrice?: number;
}

// 更新：产品列表响应可能包含制造商名称
export interface ProductListResponse {
  products: Product[]; // Product 类型已更新
  total: number;
  page: number;
  size: number;
}

// 新增：生产商列表响应接口
interface ManufacturerListResponse {
  manufacturers: Manufacturer[];
  total: number;
}

// 新增：配料列表响应接口
interface IngredientListResponse {
  ingredients: Ingredient[];
  total: number;
}

// 新增：快速创建生产商的输入数据接口
export interface ManufacturerInputData {
  name: string;
  contact_phone?: string;
  address?: string;
  certification_number?: string;
}

// 新增：快速创建配料的输入数据接口
export interface IngredientInputData {
  name: string;
  category: string;
  is_allergen: boolean;
  description?: string;
}

export const productService = {
  // 获取产品列表
  async listProducts(options: ProductListOptions): Promise<ProductListResponse> {
    const params = new URLSearchParams();
    if (options.page) params.append('page', options.page.toString());
    if (options.size) params.append('size', options.size.toString());
    if (options.category) params.append('category', options.category);
    if (options.status) params.append('status', options.status);
    if (options.keyword) params.append('keyword', options.keyword);
    if (options.sortBy) params.append('sort_by', options.sortBy);
    if (options.order) params.append('order', options.order);
    if (options.minPrice) params.append('min_price', options.minPrice.toString());
    if (options.maxPrice) params.append('max_price', options.maxPrice.toString());

    const response = await axios.get(`${API_URLS.PRODUCT.LIST}?${params.toString()}`);
    // 注意：后端返回的 product DTO 可能需要调整以匹配前端 Product 接口，尤其是 manufacturerName
    return response.data;
  },

  // 获取产品详情 (预期返回包含 manufacturer 和 ingredients)
  async getProduct(id: string): Promise<Product> {
    const response = await axios.get(API_URLS.PRODUCT.DETAIL(id));
    return response.data; // 假设后端直接返回符合更新后 Product 接口的数据
  },

  // 创建产品 (使用新的输入数据结构)
  async createProduct(data: ProductInputData): Promise<Product> {
    const response = await axios.post(API_URLS.PRODUCT.CREATE, data);
    return response.data; // 返回创建后的产品信息 (可能不含关联对象)
  },

  // 更新产品 (使用新的输入数据结构)
  async updateProduct(id: string, data: Partial<ProductInputData>): Promise<Product> {
    const response = await axios.put(API_URLS.PRODUCT.UPDATE(id), data);
    return response.data; // 返回更新后的产品信息 (可能不含关联对象)
  },

  async deleteProduct(id: string): Promise<void> {
    await axios.delete(API_URLS.PRODUCT.DELETE(id));
  },

  async getProductsByStatus(status: string, page: number = 1, size: number = 10): Promise<ProductListResponse> {
    const response = await axios.get(`${API_URLS.PRODUCT.BY_STATUS}?status=${status}&page=${page}&size=${size}`);
    return response.data;
  },

  async getProductsByBlindBox(boxId: string): Promise<Product[]> {
    const response = await axios.get(API_URLS.PRODUCT.BY_BOX(boxId));
    return response.data.products;
  },

  async addProductToBlindBox(productId: string, boxId: string): Promise<void> {
    await axios.post(API_URLS.PRODUCT.ADD_TO_BOX(productId), { blind_box_id: boxId });
  },

  async removeProductFromBlindBox(productId: string): Promise<void> {
    await axios.post(API_URLS.PRODUCT.REMOVE_FROM_BOX(productId));
  },

  async batchAddProductsToBlindBox(productIds: string[], boxId: string): Promise<void> {
    await axios.post(API_URLS.PRODUCT.BATCH_ADD_TO_BOX, {
      product_ids: productIds,
      blind_box_id: boxId,
    });
  },

  async listManufacturers(): Promise<Manufacturer[]> {
    try {
      const response = await axios.get<ManufacturerListResponse>(API_URLS.MANUFACTURER.LIST);
      return response.data.manufacturers || [];
    } catch (error) {
      console.error("Failed to fetch manufacturers:", error);
      throw error;
    }
  },

  async listIngredients(): Promise<Ingredient[]> {
    try {
      const response = await axios.get<IngredientListResponse>(API_URLS.INGREDIENT.LIST);
      return response.data.ingredients || [];
    } catch (error) {
      console.error("Failed to fetch ingredients:", error);
      throw error;
    }
  },

  async quickCreateManufacturer(data: ManufacturerInputData): Promise<Manufacturer> {
    try {
      const response = await apiClient.post<{ message: string; manufacturer_id: string }>(
        API_URLS.MANUFACTURER.CREATE,
        data
      );
      const manufacturer = await apiClient.get<Manufacturer>(
        API_URLS.MANUFACTURER.DETAIL(response.data.manufacturer_id)
      );
      return manufacturer.data;
    } catch (error) {
      console.error("Failed to create manufacturer:", error);
      throw error;
    }
  },

  async quickCreateIngredient(data: IngredientInputData): Promise<Ingredient> {
    try {
      const response = await apiClient.post<{ message: string; ingredient_id: string }>(
        API_URLS.INGREDIENT.CREATE,
        data
      );
      // 创建成功后立即获取详情
      const ingredient = await apiClient.get<Ingredient>(
        API_URLS.INGREDIENT.DETAIL(response.data.ingredient_id)
      );
      return ingredient.data;
    } catch (error) {
      console.error("Failed to create ingredient:", error);
      throw error;
    }
  }
};

File: internal/api/rest/handler/product.go
package handler

import (
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
	"go.uber.org/zap"

	"FreshBox/internal/model"
	"FreshBox/internal/service"
)

// ProductHandler 商品处理器
type ProductHandler struct {
	productService *service.ProductService
	logger         *zap.Logger
}

// NewProductHandler 创建商品处理器
func NewProductHandler(productService *service.ProductService) *ProductHandler {
	logger, _ := zap.NewDevelopment()
	return &ProductHandler{
		productService: productService,
		logger:         logger,
	}
}

// CreateProduct 创建商品
func (h *ProductHandler) CreateProduct(c *gin.Context) {
	var product model.Product
	if err := c.ShouldBindJSON(&product); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "参数错误", "details": err.Error()})
		return
	}

	// 设置创建者ID（从认证中获取）
	userID := c.GetString("user_id")
	if userID == "" {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "未授权"})
		return
	}
	product.CreatorID = userID

	err := h.productService.CreateProduct(c.Request.Context(), &product)
	if err != nil {
		h.logger.Error("创建商品失败", zap.Error(err))
		c.JSON(http.StatusInternalServerError, gin.H{"error": "创建商品失败", "details": err.Error()})
		return
	}

	c.JSON(http.StatusCreated, gin.H{
		"message":    "创建商品成功",
		"product_id": product.ID,
	})
}

// GetProduct 获取商品详情
func (h *ProductHandler) GetProduct(c *gin.Context) {
	id := c.Param("id")
	if id == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "商品ID不能为空"})
		return
	}

	product, err := h.productService.GetProduct(c.Request.Context(), id)
	if err != nil {
		h.logger.Error("获取商品失败", zap.Error(err), zap.String("product_id", id))
		c.JSON(http.StatusNotFound, gin.H{"error": "获取商品失败", "details": err.Error()})
		return
	}

	// 构建响应DTO
	productDTO := model.ProductDTO{
		ID:             product.ID,
		Name:           product.Name,
		Description:    product.Description,
		Price:          product.Price,
		Category:       product.Category,
		ImageURL:       product.ImageURL,
		Status:         product.Status,
		ProductionDate: service.FormatTime(product.ProductionDate),
		ShelfLifeHours: product.ShelfLifeHours,
		CreatedAt:      service.FormatTime(product.CreatedAt),
	}

	c.JSON(http.StatusOK, productDTO)
}

// ListProducts 列出商品
func (h *ProductHandler) ListProducts(c *gin.Context) {
	// 解析查询参数
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	size, _ := strconv.Atoi(c.DefaultQuery("size", "10"))
	category := c.Query("category")
	status := c.Query("status")
	keyword := c.Query("keyword")
	sortBy := c.Query("sort_by")
	order := c.Query("order")
	minPrice, _ := strconv.ParseFloat(c.Query("min_price"), 64)
	maxPrice, _ := strconv.ParseFloat(c.Query("max_price"), 64)

	// 构建查询选项
	opts := model.ProductListOptions{
		Page:     page,
		Size:     size,
		Category: category,
		Status:   status,
		Keyword:  keyword,
		SortBy:   sortBy,
		Order:    order,
		MinPrice: minPrice,
		MaxPrice: maxPrice,
	}

	// 查询商品列表
	products, total, err := h.productService.ListProducts(c.Request.Context(), opts)
	if err != nil {
		h.logger.Error("查询商品列表失败", zap.Error(err))
		c.JSON(http.StatusInternalServerError, gin.H{"error": "查询商品列表失败", "details": err.Error()})
		return
	}

	// 构建响应DTO
	var productDTOs []model.ProductDTO
	for _, product := range products {
		productDTOs = append(productDTOs, model.ProductDTO{
			ID:             product.ID,
			Name:           product.Name,
			Description:    product.Description,
			Price:          product.Price,
			Category:       product.Category,
			ImageURL:       product.ImageURL,
			Status:         product.Status,
			ProductionDate: service.FormatTime(product.ProductionDate),
			ShelfLifeHours: product.ShelfLifeHours,
			CreatedAt:      service.FormatTime(product.CreatedAt),
		})
	}

	c.JSON(http.StatusOK, gin.H{
		"products": productDTOs,
		"total":    total,
		"page":     page,
		"size":     size,
	})
}

// UpdateProduct 更新商品
func (h *ProductHandler) UpdateProduct(c *gin.Context) {
	id := c.Param("id")
	if id == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "商品ID不能为空"})
		return
	}

	var product model.Product
	if err := c.ShouldBindJSON(&product); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "参数错误", "details": err.Error()})
		return
	}

	// 设置ID
	product.ID = id

	err := h.productService.UpdateProduct(c.Request.Context(), id, &product)
	if err != nil {
		h.logger.Error("更新商品失败", zap.Error(err), zap.String("product_id", id))
		c.JSON(http.StatusInternalServerError, gin.H{"error": "更新商品失败", "details": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message":    "更新商品成功",
		"product_id": id,
	})
}

// DeleteProduct 删除商品
func (h *ProductHandler) DeleteProduct(c *gin.Context) {
	id := c.Param("id")
	if id == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "商品ID不能为空"})
		return
	}

	err := h.productService.DeleteProduct(c.Request.Context(), id)
	if err != nil {
		h.logger.Error("删除商品失败", zap.Error(err), zap.String("product_id", id))
		c.JSON(http.StatusInternalServerError, gin.H{"error": "删除商品失败", "details": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message":    "删除商品成功",
		"product_id": id,
	})
}

// AddProductToBlindBox 将商品添加到盲盒
func (h *ProductHandler) AddProductToBlindBox(c *gin.Context) {
	productID := c.Param("id")
	if productID == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "商品ID不能为空"})
		return
	}

	var req struct {
		BlindBoxID string `json:"blind_box_id" binding:"required"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "参数错误", "details": err.Error()})
		return
	}

	err := h.productService.AddProductToBlindBox(c.Request.Context(), productID, req.BlindBoxID)
	if err != nil {
		h.logger.Error("添加商品到盲盒失败", zap.Error(err),
			zap.String("product_id", productID),
			zap.String("blind_box_id", req.BlindBoxID))
		c.JSON(http.StatusInternalServerError, gin.H{"error": "添加商品到盲盒失败", "details": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message":      "添加商品到盲盒成功",
		"product_id":   productID,
		"blind_box_id": req.BlindBoxID,
	})
}

// RemoveProductFromBlindBox 从盲盒中移除商品
func (h *ProductHandler) RemoveProductFromBlindBox(c *gin.Context) {
	productID := c.Param("id")
	if productID == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "商品ID不能为空"})
		return
	}

	err := h.productService.RemoveProductFromBlindBox(c.Request.Context(), productID)
	if err != nil {
		h.logger.Error("从盲盒中移除商品失败", zap.Error(err), zap.String("product_id", productID))
		c.JSON(http.StatusInternalServerError, gin.H{"error": "从盲盒中移除商品失败", "details": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message":    "从盲盒中移除商品成功",
		"product_id": productID,
	})
}

// BatchAddProductsToBlindBox 批量将商品添加到盲盒
func (h *ProductHandler) BatchAddProductsToBlindBox(c *gin.Context) {
	var req struct {
		ProductIDs []string `json:"product_ids" binding:"required"`
		BlindBoxID string   `json:"blind_box_id" binding:"required"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "参数错误", "details": err.Error()})
		return
	}

	err := h.productService.BatchAddProductsToBlindBox(c.Request.Context(), req.ProductIDs, req.BlindBoxID)
	if err != nil {
		h.logger.Error("批量添加商品到盲盒失败", zap.Error(err),
			zap.Strings("product_ids", req.ProductIDs),
			zap.String("blind_box_id", req.BlindBoxID))
		c.JSON(http.StatusInternalServerError, gin.H{"error": "批量添加商品到盲盒失败", "details": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message":       "批量添加商品到盲盒成功",
		"product_count": len(req.ProductIDs),
		"blind_box_id":  req.BlindBoxID,
	})
}

// GetProductsByStatus 根据状态获取商品列表
func (h *ProductHandler) GetProductsByStatus(c *gin.Context) {
	status := c.Query("status")
	if status == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "状态参数不能为空"})
		return
	}

	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	size, _ := strconv.Atoi(c.DefaultQuery("size", "10"))

	products, total, err := h.productService.GetProductsByStatus(c.Request.Context(), status, page, size)
	if err != nil {
		h.logger.Error("查询商品列表失败", zap.Error(err), zap.String("status", status))
		c.JSON(http.StatusInternalServerError, gin.H{"error": "查询商品列表失败", "details": err.Error()})
		return
	}

	// 构建响应DTO
	var productDTOs []model.ProductDTO
	for _, product := range products {
		productDTOs = append(productDTOs, model.ProductDTO{
			ID:             product.ID,
			Name:           product.Name,
			Description:    product.Description,
			Price:          product.Price,
			Category:       product.Category,
			ImageURL:       product.ImageURL,
			Status:         product.Status,
			ProductionDate: service.FormatTime(product.ProductionDate),
			ShelfLifeHours: product.ShelfLifeHours,
			CreatedAt:      service.FormatTime(product.CreatedAt),
		})
	}

	c.JSON(http.StatusOK, gin.H{
		"products": productDTOs,
		"total":    total,
		"page":     page,
		"size":     size,
		"status":   status,
	})
}

// GetProductsByBlindBox 获取盲盒内的商品
func (h *ProductHandler) GetProductsByBlindBox(c *gin.Context) {
	boxID := c.Param("box_id")
	if boxID == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "盲盒ID不能为空"})
		return
	}

	products, err := h.productService.GetProductsByBlindBox(c.Request.Context(), boxID)
	if err != nil {
		h.logger.Error("获取盲盒商品失败", zap.Error(err), zap.String("box_id", boxID))
		c.JSON(http.StatusInternalServerError, gin.H{"error": "获取盲盒商品失败", "details": err.Error()})
		return
	}

	// 构建响应DTO
	var productDTOs []model.ProductDTO
	for _, product := range products {
		productDTOs = append(productDTOs, model.ProductDTO{
			ID:             product.ID,
			Name:           product.Name,
			Description:    product.Description,
			Price:          product.Price,
			Category:       product.Category,
			ImageURL:       product.ImageURL,
			Status:         product.Status,
			ProductionDate: service.FormatTime(product.ProductionDate),
			ShelfLifeHours: product.ShelfLifeHours,
			CreatedAt:      service.FormatTime(product.CreatedAt),
		})
	}

	c.JSON(http.StatusOK, gin.H{
		"products": productDTOs,
		"box_id":   boxID,
		"total":    len(productDTOs),
	})
}
